ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,859 @@
+    );
+
+    const { tokenAAmount, tokenBAmount } = await this.getPoolInfo();
+    const { vaultAReserve, vaultBReserve } = await this.getSplInfo();
+    const [outTotalAmount, outReserveBalance] = tokenMint.equals(this.poolState.tokenAMint)
+      ? [tokenAAmount, vaultAReserve]
+      : [tokenBAmount, vaultBReserve];
+
+    return outTotalAmount.gt(outReserveBalance) ? outReserveBalance : outTotalAmount;
+  }
+
+  public async swap(owner: PublicKey, inTokenMint: PublicKey, inAmountLamport: BN, outAmountLamport: BN) {
+    const [sourceToken, destinationToken] =
+      this.tokenA.address === inTokenMint.toBase58()
+        ? [this.poolState.tokenAMint, this.poolState.tokenBMint]
+        : [this.poolState.tokenBMint, this.poolState.tokenAMint];
+
+    const adminTokenFee =
+      this.tokenA.address === inTokenMint.toBase58() ? this.poolState.adminTokenAFee : this.poolState.adminTokenBFee;
+
+    let preInstructions: Array<TransactionInstruction> = [];
+    const [[userSourceToken, createUserSourceIx], [userDestinationToken, createUserDestinationIx]] =
+      await this.createATAPreInstructions(owner, [sourceToken, destinationToken]);
+
+    createUserSourceIx && preInstructions.push(createUserSourceIx);
+    createUserDestinationIx && preInstructions.push(createUserDestinationIx);
+
+    if (sourceToken.equals(WRAPPED_SOL_MINT)) {
+      preInstructions = preInstructions.concat(wrapSOLInstruction(owner, userSourceToken, inAmountLamport.toNumber()));
+    }
+
+    preInstructions.push(await this.getApySyncInstructions());
+
+    const postInstructions: Array<TransactionInstruction> = [];
+    if (WRAPPED_SOL_MINT.equals(destinationToken)) {
+      const unwrapSOLIx = await unwrapSOLInstruction(owner);
+      unwrapSOLIx && postInstructions.push(unwrapSOLIx);
+    }
+
+    const swapTx = await this.program.methods
+      .swap(inAmountLamport, outAmountLamport)
+      .accounts({
+        aTokenVault: this.vaultA.vaultState.tokenVault,
+        bTokenVault: this.vaultB.vaultState.tokenVault,
+        aVault: this.poolState.aVault,
+        bVault: this.poolState.bVault,
+        aVaultLp: this.poolState.aVaultLp,
+        bVaultLp: this.poolState.bVaultLp,
+        aVaultLpMint: this.vaultA.vaultState.lpMint,
+        bVaultLpMint: this.vaultB.vaultState.lpMint,
+        userSourceToken,
+        userDestinationToken,
+        user: owner,
+        adminTokenFee,
+        pool: this.address,
+        tokenProgram: TOKEN_PROGRAM_ID,
+        vaultProgram: this.vaultProgram.programId,
+      })
+      .remainingAccounts(getRemainingAccounts(this.poolState))
+      .preInstructions(preInstructions)
+      .postInstructions(postInstructions)
+      .transaction();
+
+    return new Transaction({
+      feePayer: owner,
+      ...(await this.program.provider.connection.getLatestBlockhash('finalized')),
+    }).add(swapTx);
+  }
+
+  public async getDepositQuote(
+    tokenAInAmount: BN,
+    tokenBInAmount: BN,
+    isImbalance = false,
```

Comment:
Under some circumstances it could be a bit misleading. Eg: If I pass 100 for `tokenAInAmount` and 100 for `tokenBInAmount`, isImbalance = false. The function will return me deposit quote for imbalance deposit, even I pass isImbalance = false